core/circuitbreaker/circuit_breaker.go

```diff
@@ -0,0 +1,623 @@
+	"github.com/alibaba/sentinel-golang/util"
+)
+
+type State int32
+
+/**
+  Circuit Breaker State Machine:
+
+                                 switch to open based on rule
+         +-----------------------------------------------------------------------+
+         |                                                                       |
+         |                                                                       v
++----------------+                   +----------------+      Probe      +----------------+
+|                |                   |                |<----------------|                |
+|                |   Probe succeed   |                |                 |                |
+|     Closed     |<------------------|    HalfOpen    |                 |      Open      |
+|                |                   |                |   Probe failed  |                |
+|                |                   |                +---------------->|                |
++----------------+                   +----------------+                 +----------------+
+*/
+const (
+	Closed State = iota
+	HalfOpen
+	Open
+)
+
+func (s *State) String() string {
+	switch s.get() {
+	case Closed:
+		return "Closed"
+	case HalfOpen:
+		return "HalfOpen"
+	case Open:
+		return "Open"
+	default:
+		return "Undefined"
+	}
+}
+
+func (s *State) get() State {
+	statePtr := (*int32)(unsafe.Pointer(s))
+	return State(atomic.LoadInt32(statePtr))
+}
+func (s *State) set(update State) {
+	statePtr := (*int32)(unsafe.Pointer(s))
+	newState := int32(update)
+	atomic.StoreInt32(statePtr, newState)
+}
+func (s *State) casState(expect State, update State) bool {
+	statePtr := (*int32)(unsafe.Pointer(s))
+	oldState := int32(expect)
+	newState := int32(update)
+	return atomic.CompareAndSwapInt32(statePtr, oldState, newState)
+}
+
+type StateChangeListener interface {
+	OnChangeToClosed(prev State, rule Rule)
+
+	OnChangeToOpen(prev State, rule Rule, snapshot interface{})
+
+	OnChangeToHalfOpen(prev State, rule Rule)
+}
+
+type CircuitBreaker interface {
+	BoundRule() Rule
+
+	BoundStat() interface{}
+
+	TryPass(ctx *base.EntryContext) bool
+
+	CurrentState() State
+	// HandleCompleted handle the entry completed
+	// rt: the response time this entry cost.
+	HandleCompleted(rt uint64, err error)
+}
+
+//================================= circuitBreakerBase ====================================
+type circuitBreakerBase struct {
+	rule               Rule
+	retryTimeoutMs     uint32
+	nextRetryTimestamp uint64
+	status             *State
+}
+
+func (b *circuitBreakerBase) BoundRule() Rule {
+	return b.rule
+}
+
+func (b *circuitBreakerBase) CurrentState() State {
+	return b.status.get()
+}
+
+// Must be overridden by concrete circuitBreaker
+func (b *circuitBreakerBase) resetMetric() {
+	logger.Warnf("Must be called by sub struct.")
+}
+
+func (b *circuitBreakerBase) retryTimeoutArrived() bool {
+	return util.CurrentTimeMillis() >= b.nextRetryTimestamp
```

Comment:
We may need to make it thread-safe when accessing and modifying  `nextRetryTimestamp`.